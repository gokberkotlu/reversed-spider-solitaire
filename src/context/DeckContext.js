import React, { useState, createContext, useEffect } from "react";
import { cardSeries } from '../utils/CardSeries';

export const DeckContext = createContext();

export const DeckProvider = props => {

    // States
    const [gameCards, setGameCards] = useState([[],[],[],[],[],[],[],[],[],[]]); // Represents cards on table
    const [remainingCards, setRemainingCards] = useState([]); // Cards outside the game
    const [numberOfCompleted, setNumberOfCompleted] = useState(Array(0)); // Shows number of completed series
    const [newCardPacks, setNewCardPacks] = useState(Array(5)); // Shows number of new card packs
    const [isGameStarted, setIsGameStarted] = useState(false); // Set state when any card action happened
    const [statesLog, setStatesLog] = useState([]); // Log States to Undo the Game

    // States: Score Datas
    const [score, setScore] = useState(500); // Game score which starts 500
    const [highScore, setHighScore] = useState(+localStorage.getItem('highScore')); // Recorded highscore to store in localStorage

    const dealTheCards = (cards, numberOfCards) => {
        
        // Save Changes When the New Card Pack Added
        logGameChanges();

        if(cards.length >= 10) {
            let gameSet = [...gameCards];
            for(let i = 0; i < numberOfCards; i++) {
                gameSet[i%10].push(cards[i]);
            }
            setGameCards(gameSet);
            setRemainingCards(cards.splice(numberOfCards));

            return gameSet;
        }
    }

    // arrangeReversedCards arranges the reversed card at the start
    const arrangeReversedCards = (gameSet) => {
        gameSet.forEach(cards => {
            for(let i = 0; i < cards.length - 1; i++) {
                cards[i].isCardReverse = true;
            }
        })
    }

    // suffleTheCards shuffles eight decks of cards
    const suffleTheCards = () => {
        let deskNumber = 8;
        let newArray = [];
        for(let i = 0; i < deskNumber; i++) {
            newArray = newArray.concat(
                [...cardSeries.map((card) => {
                    return {
                        "cardNum": card,
                        "cardDeck": (i + 1),
                        "isCardReverse": false
                    }
                })]);
        }
        var currentIndex = newArray.length,
            randomIndex;
        // While there remain elements to shuffle...
        while (0 !== currentIndex) {
            // Pick a remaining element...
            randomIndex = Math.floor(Math.random() * currentIndex);
            currentIndex--;
    
            // And swap it with the current element.
            [newArray[currentIndex], newArray[randomIndex]] = [
                newArray[randomIndex], newArray[currentIndex]
            ];
        }
        return newArray;
    }

    // logGameChanges takes snapshots before all card actions and stores them inside statesLog
    const logGameChanges = () => {
        let addToLog = {
            "gameCardsBefore": JSON.parse(JSON.stringify(gameCards)),
            "remainingCardsBefore": JSON.parse(JSON.stringify(remainingCards)),
            "numberOfCompletedBefore": JSON.parse(JSON.stringify(numberOfCompleted)),
            "newCardPacksBefore": JSON.parse(JSON.stringify(newCardPacks))
        }
        setStatesLog(prevState => [...prevState, addToLog]);
    }

    // undoTheGame sets game states with last logged snapshot
    const undoTheGame = () => {
        let statesLogValue = [...statesLog];
        let statesLogValueLength = statesLogValue.length
        if(statesLogValueLength > 1) {
            let beforeStates = statesLogValue[statesLogValueLength - 1]; // Go back in the game by having "before value of last screen"
            setGameCards(beforeStates.gameCardsBefore);
            setRemainingCards(beforeStates.remainingCardsBefore);
            setNumberOfCompleted(beforeStates.numberOfCompletedBefore);
            setNewCardPacks(beforeStates.newCardPacksBefore);

            // Delete current states from log when went back in the game
            setStatesLog(statesLogValue.slice(0, statesLogValueLength - 1));

            // Decrease 1 point from score for each undo click
            setScore(prevState => prevState - 1);

            // Decrease 100 points from score if completed desk is removed by clicking undo button
            if(beforeStates.numberOfCompletedBefore.length < numberOfCompleted.length) {
                let scoreDiff = numberOfCompleted.length - beforeStates.numberOfCompletedBefore.length;
                setScore(prevState => prevState - 100 * scoreDiff);
            }
        }
    }

    useEffect(() => {
        // Shuffle cards, arrange reversed cards and deal them on the game table at start
        var allCards = suffleTheCards();
        var gameSet = dealTheCards(allCards, 54);
        arrangeReversedCards(gameSet);
    }, [])

    return (
        <DeckContext.Provider
        value={{
                "gameCards": [gameCards, setGameCards],
                "remainingCards": [remainingCards, setRemainingCards],
                "numberOfCompleted": [numberOfCompleted, setNumberOfCompleted],
                "newCardPacks": [newCardPacks, setNewCardPacks],
                "isGameStarted": [isGameStarted, setIsGameStarted],
                "suffleTheCards": suffleTheCards,
                "dealTheCards": dealTheCards,
                "arrangeReversedCards": arrangeReversedCards,
                "statesLog": [statesLog, setStatesLog],
                "logGameChanges": logGameChanges,
                "undoTheGame": undoTheGame,
                "score": [score, setScore],
                "highScore": [highScore, setHighScore]
            }}>
            {props.children}
        </DeckContext.Provider>
    );
}