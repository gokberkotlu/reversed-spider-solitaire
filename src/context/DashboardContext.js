import React, { useState, createContext, useContext, useEffect } from "react";
import { DeckContext } from "./DeckContext";

export const DashboardContext = createContext();

export const DashboardProvider = props => {

    // States
    const [counter, setCounter] = useState("00:00:00"); // Represents timer
    const [isGameFinished, setIsGameFinished] = useState(false); // According to this state, popup is shown or not
    const [resetBit, setResetBit] = useState(false); // It triggers the useEffect to reset the game
    const [decreaseInEveryTenSeconds, setDecreaseInEveryTenSeconds] = useState(0); // Decrease 2 points from score for each 10 sec

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [, setGameCards] = deckContextValue.gameCards;
    const [, setRemainingCards] = deckContextValue.remainingCards;
    const [, setNumberOfCompleted] = deckContextValue.numberOfCompleted;
    const [, setNewCardPacks] = deckContextValue.newCardPacks;
    const [, setIsGameStarted] = deckContextValue.isGameStarted;
    const [, setStatesLog] = deckContextValue.statesLog;
    const suffleTheCards = deckContextValue.suffleTheCards;
    const dealTheCards = deckContextValue.dealTheCards;
    const arrangeReversedCards = deckContextValue.arrangeReversedCards;

    const [, setScore] = deckContextValue.score;

    const resetGame = () => {
        // Resetting
        setGameCards([[],[],[],[],[],[],[],[],[],[]]);
        setRemainingCards([]);
        setNewCardPacks(Array(5));
        setNumberOfCompleted(Array(0));
        setScore(500);
        setCounter("00:00:00");
        setIsGameFinished(false);
        setResetBit(true);
        setIsGameStarted(false);
        setStatesLog([]);
        setDecreaseInEveryTenSeconds(0);
    }

    useEffect(() => {// Deal cards again from scratch
        if(resetBit) {
            var allCards = suffleTheCards();
            var gameSet = dealTheCards(allCards, 54); // setGameCards and setRemainingCards included
            arrangeReversedCards(gameSet);
            setResetBit(false);
        }
    }, [resetBit]) // resetBit changes trigger the control of gameSet and remainingCards

    return (
        <DashboardContext.Provider
        value={{"counter": [counter, setCounter],
        "isGameFinished": [isGameFinished, setIsGameFinished],
        "resetGame": resetGame,
        "decreaseInEveryTenSeconds": [decreaseInEveryTenSeconds ,setDecreaseInEveryTenSeconds]}}>
            {props.children}
        </DashboardContext.Provider>
    );
}