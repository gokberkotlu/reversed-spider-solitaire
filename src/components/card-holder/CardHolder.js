import React, { useContext, useEffect } from 'react';
import Card from '../card/Card';
import EmptyCardBox from '../empty-card-box/EmptyCardBox';
import { DeckContext } from '../../context/DeckContext';
import { DashboardContext } from '../../context/DashboardContext';
import { cardSeries } from '../../utils/CardSeries';
import './card-holder.css';

const CardHolder = ({cards, cardUnitNumber}) => {

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [gameCards, setGameCards] = deckContextValue.gameCards;
    const [numberOfCompleted, setNumberOfCompleted] = deckContextValue.numberOfCompleted;
    const [isGameStarted, setIsGameStarted] = deckContextValue.isGameStarted;
    const logGameChanges = deckContextValue.logGameChanges;
    const [score, setScore] = deckContextValue.score;
    const [highScore, setHighScore] = deckContextValue.highScore;

    // DashboardContext
    const dashboardContextValue = useContext(DashboardContext);
    
    const [, setIsGameFinished] = dashboardContextValue.isGameFinished;


    const allowDrop = (event) => {
        event.preventDefault();
    }

    const addSelectedCards = (ids, prevCards, beforeUnitIndex, currentUnitIndex) => {
        // Push selected cards to prevCard state in valid unit
        for(let i = 0; i < ids.length; i++) {
            let movingCardNumber = ids[i].split("-")[0];
            let deckNumberOfMovingCard = +ids[i].split("-")[4];
            prevCards[beforeUnitIndex].map((card, index) => {
                // Prevent this stages if card is droping to same unit
                if(card.cardNum === movingCardNumber && card.cardDeck === deckNumberOfMovingCard) {
                    var movingCard = prevCards[beforeUnitIndex][index];
                    // Delete replaced cards from before unit
                    prevCards[beforeUnitIndex].splice(index, 1);
                    // Add selected cards to new unit
                    prevCards[currentUnitIndex].push(movingCard);
                }
                return prevCards;
            });
        }
        setGameCards(prevCards);
    }
    
    const drop = (event) => {

        // Save Changes Made by Drop Event
        logGameChanges();

        // set game cards with getData
        var data = event.dataTransfer.getData("Text");
        // If there are movable cards
        if(data.length !== 0) {
            var currentUnitIndex = cardUnitNumber - 1;
            var ids = data.split(",");
            var firstCardDragging = ids[0].split("-")[0];
            var beforeUnitIndex = +ids[0].split("-")[2] - 1;
            // If card unit is changing for the valid card/cards
            if(currentUnitIndex !== beforeUnitIndex) {

                var prevCards = [...gameCards];
                // Get the last card's number in current unit and check if suitable
                // Also check if the current unit has any cards
                let currentUnit = [...prevCards[currentUnitIndex]];
                if(currentUnit.length !== 0) {
                    let lastCardBeforeDrop = currentUnit.slice(-1)[0].cardNum;
                    if(cardSeries.indexOf(lastCardBeforeDrop) === cardSeries.indexOf(firstCardDragging) - 1) {
                        addSelectedCards(ids, prevCards, beforeUnitIndex, currentUnitIndex);
                        // Start timer when first 10 card pack added
                        if(!isGameStarted) {
                            setIsGameStarted(true);
                        }
                        // Decrease score with one when a move realized
                        setScore(prevScore => prevScore - 1);
                    }
                } else {
                    addSelectedCards(ids, prevCards, beforeUnitIndex, currentUnitIndex)
                }
            }
        }
    }

    const openLastCardIfReversed = () => {
        // This query turns the card if it's the last reversed card of its unit
        if(cards.length !== 0 && cards[cards.length - 1].isCardReverse) {
            let cloneGameCards = [...gameCards];
            cloneGameCards[cardUnitNumber - 1][cards.length - 1].isCardReverse = false;
            setGameCards(cloneGameCards);
        }
    }

    const isSeriesCompleted = async () => {
        if(cards.length >= 13 && cards.slice(-1)[0].cardNum === 'K') {
            let lastCardIndex = cards.length - 1;
            let compareRepeat = 0;
            // Limiter : 12 -> To get first card of series by (index - 1)
            for(let i = lastCardIndex; i > lastCardIndex - 12; i--) {
                if((cardSeries.indexOf(cards[i].cardNum) === cardSeries.indexOf(cards[i - 1].cardNum) + 1)
                    &&
                    !cards[i - 1].isCardReverse
                ) {
                    compareRepeat += 1; // Keep data of how many comparing steps are done
                    if(compareRepeat === 12) {

                        let cloneCards = [...cards];
                        let cloneGameCards = [...gameCards];

                        cloneCards.splice(i - 1,); // Delete the complete series from gameCards
                        cloneGameCards[cardUnitNumber - 1] = cloneCards;
                        await setGameCards(cloneGameCards);

                        setNumberOfCompleted(Array(numberOfCompleted.length + 1)); // Increase completed desk blocks

                        // Check if game finished
                        if(numberOfCompleted.length + 1 === 8) { // 8 stands for completed decks
                            setIsGameFinished(true); // Get the pop-up with this dependency
                            setIsGameStarted(false); // Stop the timer which is bounded to this dependency when game is completed
                        }

                        // Score states setting
                        let updatedScore = score + 100; // Update score to set state of score and highScore

                        setScore(updatedScore); // Increase score, if a series completed
                        if(updatedScore > highScore) { // Update highScore if there is new
                            localStorage.setItem('highScore', updatedScore);
                            setHighScore(updatedScore);
                        }
                    }
                } else {
                    break;
                } // if statement
            } // for statement
        } // if statement
    } // isSeriesCompleted

    useEffect(() => {
        isSeriesCompleted(); // Check if a card series
        openLastCardIfReversed(); // Check if last card is reversed in it's unit
    }, [gameCards])

    return (
        <div
        className="card-holder-unit"
        onDrop={(event) => drop(event)}
        onDragOver={(event) => allowDrop(event)}>
            <EmptyCardBox />
            {/* cards variable contains the cards inside card unit(including column) */}
            {cards.map((card) => (
                <Card
                cardNumber={card.cardNum}
                cardUnitNumber={cardUnitNumber}
                cardDeck={card.cardDeck}
                isCardReverse={card.isCardReverse}
                key={'rank-' + card.cardNum + '-deck-' + card.cardDeck}/>
            ))}
        </div>
    );
}

export default React.memo(CardHolder);