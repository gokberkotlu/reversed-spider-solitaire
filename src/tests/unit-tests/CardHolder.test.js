import renderer from "react-test-renderer";
import CardHolder from "../../components/card-holder/CardHolder";
import { DeckProvider } from '../../context/DeckContext';
import { DashboardProvider } from "../../context/DashboardContext"

const cards = [
    {
        "cardNum": 'A',
        "cardDeck": 1,
        "isCardReverse": true
    },
    {
        "cardNum": '2',
        "cardDeck": 1,
        "isCardReverse": true
    },
    {
        "cardNum": '3',
        "cardDeck": 1,
        "isCardReverse": false
    },
    {
        "cardNum": '4',
        "cardDeck": 1,
        "isCardReverse": false
    }
]
const cardUnitNumber = 1;

describe('CardHolder component', () => {
    it('should render CardHolder component', () => {
        const tree = renderer.create(
            <DeckProvider value={null}>
                <DashboardProvider value={null}>
                    <CardHolder cards={cards} cardUnitNumber={cardUnitNumber}/>
                </DashboardProvider>
            </DeckProvider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })
})