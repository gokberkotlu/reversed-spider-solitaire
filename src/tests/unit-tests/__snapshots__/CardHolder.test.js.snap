// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardHolder component should render CardHolder component 1`] = `
<div
  className="card-holder-unit"
  onDragOver={[Function]}
  onDrop={[Function]}
>
  <div
    className="empty-card-box"
  />
  <div
    className="card"
    draggable={true}
    id="A-spare-1-deck-1"
    onDrag={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="card-template card-spade card-back"
    />
  </div>
  <div
    className="card"
    draggable={true}
    id="2-spare-1-deck-1"
    onDrag={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="card-template card-spade card-back"
    />
  </div>
  <div
    className="card"
    draggable={true}
    id="3-spare-1-deck-1"
    onDrag={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="card-template card-spade card-front"
    >
      <div>
        <div
          className="card-rank"
        >
          3
        </div>
        <div
          className="spade-icon"
        >
          ♠
        </div>
      </div>
      <div
        className="spade-center"
      >
        ♠
      </div>
      <div>
        <div
          className="spade-icon reversed-icon"
        >
          ♠
        </div>
        <div
          className="card-rank reversed-icon"
        >
          3
        </div>
      </div>
    </div>
  </div>
  <div
    className="card"
    draggable={true}
    id="4-spare-1-deck-1"
    onDrag={[Function]}
    onDragEnd={[Function]}
    onDragStart={[Function]}
  >
    <div
      className="card-template card-spade card-front"
    >
      <div>
        <div
          className="card-rank"
        >
          4
        </div>
        <div
          className="spade-icon"
        >
          ♠
        </div>
      </div>
      <div
        className="spade-center"
      >
        ♠
      </div>
      <div>
        <div
          className="spade-icon reversed-icon"
        >
          ♠
        </div>
        <div
          className="card-rank reversed-icon"
        >
          4
        </div>
      </div>
    </div>
  </div>
</div>
`;
